bootstrap/tracing.go

```diff
@@ -0,0 +1,155 @@
+package bootstrap
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/buildkite/agent/v3/agent"
+	"github.com/buildkite/agent/v3/env"
+	"github.com/buildkite/agent/v3/tracetools"
+	"github.com/opentracing/opentracing-go"
+	"golang.org/x/exp/slices"
+	ddext "gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/opentracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+// stopper lets us abstract the tracer wrap up code so we can plug in different tracing
+// library implementations that are opentracing compatible. Opentracing itself
+// doesn't have a Stop function on its Tracer interface.
+type stopper func()
+
+func noopStopper() {}
+
+func (b *Bootstrap) startTracing(ctx context.Context) (tracetools.Span, context.Context, stopper) {
+	switch b.Config.TracingBackend {
+	case tracetools.BackendDatadog:
+		// Newer versions of the tracing libs print out diagnostic info which spams the
+		// Buildkite agent logs. Disable it by default unless it's been explicitly set.
+		if _, has := os.LookupEnv("DD_TRACE_STARTUP_LOGS"); !has {
+			os.Setenv("DD_TRACE_STARTUP_LOGS", "false")
+		}
+
+		return b.startTracingDatadog(ctx)
+
+	case tracetools.BackendNone:
+		return &tracetools.NoopSpan{}, ctx, noopStopper
+
+	default:
+		b.shell.Commentf("An invalid tracing backend was provided: %q. Tracing will not occur.", b.Config.TracingBackend)
+		b.Config.TracingBackend = "" // Ensure that we don't do any tracing after this, some of the stuff in tracetools uses the bootstrap's tracking backend
```

Comment:
this step was necessary in previous implementations, because tracetools.StartSpanFromContext panicked when passed an invalid tracing backend, and this would stop that from happening. It's technically no longer necessary, but i still kinda like it from a defensive programming perspective. What do you think?

Reply:
ooh yes, good call!